[BT] Add `DPT` support for `BetterTransformer`

hazrulakmal opened the pull request:
# What does this PR do?

Fixes # [20372](https://github.com/huggingface/transformers/issues/20372)

1. Add a new class `DPTViTLayerBetterTransformer`. Make ViTLayer as the main reference for forward pass
2. update the BETTER_TRANFORMER_LAYERS_MAPPING_DICT dictionary in optimum/bettertransformer/__init__.py
3. Add `hf-internal-testing/tiny-random-DPTModel`
4. Run `pytest tests/bettertransformer//test_bettertransformer_vision.py`
Got some errors but they are not related to DPTViTLayer. Below was what I got rather
>  ============== short test summary info ================
FAILED tests/bettertransformer/test_bettertransformer_vision.py::BetterTransformersVisionTest::test_logits - TypeError: 'str' object cannot be interpreted as an integer
FAILED tests/bettertransformer/test_bettertransformer_vision.py::BetterTransformersVisionTest::test_raise_autocast - TypeError: 'str' object cannot be interpreted as an integer
FAILED tests/bettertransformer/test_bettertransformer_vision.py::BetterTransformersVisionTest::test_raise_train - TypeError: 'str' object cannot be interpreted as an integer
FAILED tests/bettertransformer/test_bettertransformer_vision.py::BetterTransformersViLTTest::test_logits - AttributeError: module 'collections' has no attribute 'MutableMapping'
FAILED tests/bettertransformer/test_bettertransformer_vision.py::BetterTransformersCLIPTest::test_logits_0_hf_internal_testing_tiny_random_clip_zero_shot_image_classification - AttributeError: module 'collections' has no attribute 'MutableMapping'
FAILED tests/bettertransformer/test_bettertransformer_vision.py::BetterTransformersCLIPTest::test_logits_1_hf_internal_testing_tiny_random_clip_zero_shot_image_classification - AttributeError: module 'collections' has no attribute 'MutableMapping'
FAILED tests/bettertransformer/test_bettertransformer_vision.py::BetterTransformersCLIPTest::test_logits_2_laion_CLIP_ViT_B_32_laion2B_s34B_b79K - AttributeError: module 'collections' has no attribute 'MutableMapping'
FAILED tests/bettertransformer/test_bettertransformer_vision.py::BetterTransformersCLIPTest::test_logits_3_laion_CLIP_ViT_B_32_laion2B_s34B_b79K - AttributeError: module 'collections' has no attribute 'MutableMapping'
FAILED tests/bettertransformer/test_bettertransformer_vision.py::BetterTransformersCLIPTest::test_raise_on_save - NotADirectoryError: [WinError 267] The directory name is invalid: 'C:\\Users\\HAZRUL~1\\AppData\\Local\\Temp\\tmpe1...

5. add a link to DPT Arxiv paper in the list of supported models [overview.mdx](https://github.com/huggingface/optimum/blob/main/docs/source/bettertransformer/overview.mdx)

To : @younesbelkada

younesbelkada commented:
Hi @hazrulakmal !
Thanks so much for the very clean PR!
Actually only 2 tests are failing, they should be fixed by adding a special condition here: https://github.com/huggingface/optimum/blob/main/tests/bettertransformer/test_bettertransformer_encoder.py#L93 as we do it for `DistilBert` etc where you return `DPT` if `layer_class=="DPTViT"`

hazrulakmal commented:
> Hi @hazrulakmal ! Thanks so much for the very clean PR! Actually only 2 tests are failing (logits tests are passing!), they should be fixed by adding a special condition here: https://github.com/huggingface/optimum/blob/main/tests/bettertransformer/test_bettertransformer_encoder.py#L93 as we do it for `DistilBert` etc where you return `DPT` if `layer_class=="DPTViT"`

But DPT is a vision model, right? shouldn't we be changing code in the `https://github.com/huggingface/optimum/blob/main/tests/bettertransformer/test_bettertransformer_vision.py` ?

younesbelkada commented:
Yes you are right, some refactoring has been done in https://github.com/huggingface/optimum/pull/594 and the test has been removed
Could you please rebase with the main branch, run the styling changes `make style` and we should be able to merge this one!

hazrulakmal commented:
I have made the necessary changes. pytest file is an empty file that I accidentally created, so had to make another commit to remove it.

hazrulakmal commented:
**[Recent update]**

I have removed unnecessary commits (there were 15 unnecessary commits due to me messing up with git just now). I have, however, cleaned the mistakes I made and reopened a fresh commit that should reflect the latest changes from the main branch.

Unfortunately, when I run `pytest tests/bettertransformer//test_bettertransformer_vision.py::BetterTransformersVisionTest`  this time around. I got three failed, and only two tests passed 

> ============================ short test summary info ===============================
FAILED tests/bettertransformer/test_bettertransformer_vision.py::BetterTransformersVisionTest::test_logits - TypeError: 'str' object cannot be interpreted as an integer
FAILED tests/bettertransformer/test_bettertransformer_vision.py::BetterTransformersVisionTest::test_raise_autocast - TypeError: 'str' object cannot be interpreted as an integer
FAILED tests/bettertransformer/test_bettertransformer_vision.py::BetterTransformersVisionTest::test_raise_train - TypeError: 'str' object cannot be interpreted 
as an integer
============================= 3 failed, 2 passed in 10.80s =============================

hazrulakmal commented:
Hi @younesbelkada & @fxmarty, I think this PR probably has gone off the radar.  😅 would like you both to review it.
